PIPE2D-83: Travis-CI

PaulPrice opened the pull request.

PaulPrice committed:
add script to build PFS pipeline

PaulPrice committed:
add script to run integration test

This will exercise the pipeline.

PaulPrice committed:
add scripts to run under Travis-CI

PaulPrice committed:
add templates for running other packages under Travis-CI

We want other packages (e.g., drp_stella, obs_pfs) to run the
pipeline under Travis, which involves checking this package out.
The travis.sh.TEMPLATE and travis.yml.TEMPLATE provide the
necessary machinery to do that. Unfortunately, they need to be
copied into the other packages (as travis.sh and .travis.yml).
The .TEMPLATE files here in this package will serve as the
authoritative version.

jdswinbank commented on .travis.yml:
```diff
@@ -0,0 +1,30 @@
+sudo: false
+python:
+  - 2.7
+#  - 3.5
```
Why leave this in-but-commented?

jdswinbank commented on .travis.yml:
```diff
@@ -0,0 +1,30 @@
+#  - 3.5
+os:
+  - linux
+#  - osx
```
Why not osx? I actually didn't realise it was possible to use OS X with Travis, but since it evidently is, why wouldn't we want to?

jdswinbank commented on .travis.yml:
```diff
@@ -0,0 +1,30 @@
+  email: false
+
+install: true
+script: ./bin/pfs_travis.sh
```
Two comments here:

1. It might be nice to add a brief comment to explain what's going on here: to the non-Travis cognoscenti, `install: true` doesn't obviously mean "skip the install step".
2. Why are we skipping the installation step anyway? Naïvely, one might assume you'd run `install_pfs.sh` as the install and `pfs_tutorial.sh` as the build (/`script` in this terminology). I assume there are good reasons for not doing so (and I might even learn them as I continue reading), but a brief explanation here would certainly help.

jdswinbank commented on .travis.yml:
```diff
@@ -0,0 +1,30 @@
+before_cache:
+  - find $HOME/.casher -name "*.tgz" -print -exec rm {} \;
+  - rm -r -f $HOME/pfs/pfs
+  - rm -r -f $HOME/pfs/tutorial
```
No newline?

jdswinbank commented on .travis.yml:
```diff
@@ -0,0 +1,30 @@
+install: true
+script: ./bin/pfs_travis.sh
+
+cache:
```
This is another place where a small comment might help. I think the aim here is to cache the directory which has LSST stack installed in it but to clobber the directory with the pfs packages (`pfs/pfs`) and that with the tutorial outputs (`pfs/tutorial`), but it's not immediately obvious why we need to prune `*.tgz` in addition to that, or what the motivation behind the timeout is. As above, I expect a lot of this will become clearer with more reading, but a brief explanation could save some head-scratching.

jdswinbank replied:
Updated following your latest commit — I see the `find` has gone, but the “bleeding edge casher” doesn't seem to be documented at https://docs.travis-ci.com/user/caching, so I think maybe a little more explanation might still be helpful.

jdswinbank commented on bin/pfs_tutorial.sh:
```diff
@@ -0,0 +1,124 @@
+cd $PREFIX
+
+# Setting lfs.batch=true enables passwordless downloads with git-lfs.
+if [ -e drp_stella_data ]; then
```
I was surprised to find this here rather than as part of `install_pfs.sh`. I guess it makes some sense if you regard the install script as covering the code that should be rebuilt on every run, while the data repository can be cached. I suppose that's fine as long as we have only a single data repository, but it feels slightly awkward to use `pfs_packages.txt` to avoid baking the names of the packages into your scripts... and then to go ahead and bake in drp_stella_data anyway.

jdswinbank commented on bin/install_pfs.sh:
```diff
@@ -0,0 +1,198 @@
+    # Clone a git repository, optionally checking out a particular version
+    local repo=$1  # Repository
+    local commit=$2  # Commit/branch to checkout
+    local required=${3:-false}  # Commit/branch is required?
```
Is the `required` argument useful? Can't you assume that the commit is required if it has been supplied?

jdswinbank replied:
On further reading: I think this relates to my comment about `BRANCH` below. *If* you were supplying `BRANCH` as an argument at line 184, this becomes necessary, but since you aren't it doesn't add anything.

jdswinbank commented on bin/pfs_travis.sh:
```diff
@@ -0,0 +1,61 @@
+# Important envvars set under Travis:
+# * TRAVIS_BRANCH: For builds not triggered by a pull request this is the name of the branch currently being
+#   built; whereas for builds triggered by a pull request this is the name of the branch targeted by the pull
+#   request (in many cases this will be master).
```
(Oh, wow, that's confusing. Bad Travis!)

jdswinbank commented on bin/install_pfs.sh:
```diff
@@ -0,0 +1,198 @@
+while getopts ":b:s" opt; do
+    case "${opt}" in
+        b)
+            BRANCH=${OPTARG}
```
Where is this used?

Actually, looking closer: I don't think this is used anywhere, and I think that's a problem. Isn't the idea that you should be using this at line 184 below to ensure you're setting up the right branches, but you currently aren't?

jdswinbank commented on bin/install_pfs.sh:
```diff
@@ -0,0 +1,198 @@
+        git checkout $commit || ( echo "Cannot checkout $commit" && ! $required )
+    fi
+    setup -k -r .
+    [ -e SConstruct ] && scons
```
Seems weird to be doing the `setup` & `scons` in a function called `git_clone`. Not a big deal, but maybe something line `clone_and_build` would be more obvious?

jdswinbank commented on bin/install_pfs.sh:
```diff
@@ -0,0 +1,198 @@
+
+# Install LSST stack
+if [ ! -e $PREFIX/conda/envs/$LSST_VERSION ]; then
+    conda config --add channels http://conda.lsst.codes/stack/0.12.1
```
Seems weird to set `LSST_VERSION` above, but then hard-code 0.12.1 here.

jdswinbank commented on bin/install_pfs.sh:
```diff
@@ -0,0 +1,198 @@
+
+mkdir -p $PREFIX/lsst
+touch $SETUPS_LSST
+[ -e $PREFIX/pfs ] && rm -rf $PREFIX/pfs  # Want to regenerate PFS-specific packages every time
```
Is this degenerate with the `cache` stanza in `.travis.yml`?

jdswinbank commented on bin/pfs_travis.sh:
```diff
@@ -0,0 +1,61 @@
+
+# Work around log length limitation (4 MB) by trapping all output.
+# In the event of an error, we'll dump the last 100 lines.
+# This code from http://stackoverflow.com/a/26082445/834250
```
(C/Sh)ould we post the full log to somewhere the user can retrieve it from?

jdswinbank commented on bin/pfs_travis.sh:
```diff
@@ -0,0 +1,61 @@
+# If an error occurs, run our error handler to output a tail of the build
+trap 'error_handler' ERR
+
+# Set up a repeating loop to send some output to Travis.
```
Is this required to prevent Travis from thinking we've timed out, or is it just a comfort display for the user?

jdswinbank commented on bin/pfs_travis.sh:
```diff
@@ -0,0 +1,61 @@
+$HERE/bin/install_pfs.sh -b $BUILD_BRANCH $HOME/pfs >> $BUILD_OUTPUT 2>&1
+. $HOME/pfs/pfs_setups.sh >> $BUILD_OUTPUT 2>&1
+git lfs install  # May not have been done if using cache
+$HERE/bin/pfs_tutorial.sh -b $BUILD_BRANCH -c 4 $HOME/pfs/tutorial >> $BUILD_OUTPUT 2>&1
```
Why `-c 4`? Do all the Travis build slaves have four cores?

...actually, according to https://docs.travis-ci.com/user/ci-environment/#Virtualization-environments they have two in this environment (Linux, `sudo: false`).

jdswinbank commented on bin/pfs_tutorial.sh:
```diff
@@ -0,0 +1,124 @@
+#
+# Exercise the PFS 2D pipeline code.
+#
+# We run through the tutorial to make sure everything's working.
```
I wonder if calling this "the tutorial" is helpful in the long term? I understand that for now it's directly derived from the “getting started” guide, but I'd imagine that long term our needs for CI and for user-focused documentation might diverge.

jdswinbank commented on pfs_packages.txt:
```diff
@@ -0,0 +1,5 @@
+# Packages should be listed in the order they will be built
+# Each line should contain the repository
+git://github.com/PaulPrice/obs_pfs
```
Is your obs_pfs special? Not sure why this doesn't point to Subaru-PFS.

jdswinbank replied:
Oh, saving space via INFRA-53?

jdswinbank commented on bin/install_pfs.sh:
```diff
@@ -0,0 +1,198 @@
+# Install PFS packages
+cd $PREFIX/pfs
+sed -e 's/#.*$//' -e '/^$/d' $HERE/pfs_packages.txt | while read REPO; do
+    git_clone $REPO
```
I think you need to be passing `$BRANCH false` here.

PaulPrice committed:
travis: use bleeding-edge casher for cache deletion

Travis have just implemented a feature I requested: deleting the
cache after it has been downloaded and extracted. Therefore I can
delete my workaround and enable use of the bleeding-edge casher.
